### ISSUE TITLE
Completing read candidates do not appear when reading from transient with customized `transient-display-buffer-action`

### ISSUE THREAD

**Issue by josephmturner** [minad/vertico#532]
Hello!  I'm not sure if it would be more appropriate to post this issue here or in the transient repository, but I'll include @tarsius in the conversation. 

If `transient-display-buffer-action` has been customized, then calls to `completing-read` from within a transient suffix (with `:transient t`, which keeps the transient open while reading input from the user) cause vertico to not show any candidates.

Tested on 29.4:

```
(require 'vertico)
(require 'transient)

(vertico-mode +1)

;; same window: no candidates displayed
(setf transient-display-buffer-action '(display-buffer-same-window))

;; side window on left, right, or top: also no candidates displayed
;; (setf transient-display-buffer-action '(display-buffer-in-side-window
;;                                         (side . right)))

;; Default: candidates displayed as expected
;; (setf transient-display-buffer-action '(display-buffer-in-side-window
;;                                         (side . bottom)
;;                                         (dedicated . t)
;;                                         (inhibit-same-window . t)))

(transient-define-prefix test-vertico-transient ()
  "Show menu for HYPERDRIVE peer graph."
  ;; TODO: Update info manual link
  :info-manual "(hyperdrive)"
  :refresh-suffixes t
  ["Testing vertico with transient display buffer action"
   ("s" "Read string" test-suffix)]
  (interactive)
  (transient-setup 'test-vertico-transient))

(transient-define-suffix test-suffix (string)
  :transient t
  (interactive
   (list (completing-read "Choose one: " '("a" "b" "c"))))
  (message "You said: %s" string))

;; (test-vertico-transient)
;; Then press `s` to choose from the list of candidates.
;; If `transient-display-buffer-action' isn't set to a bottom side window,
;; `completing-read' candidates don't appear.
```

Thank you!

Joseph

**Comment by minad** [comment]
Hi. I can confirm this problem, but haven't figured out a solution. Vertico properly calls `window-resize` in `vertico--resize-window`, but the resize call is ineffective. It works however if I set `transient-hide-during-minibuffer-read=t`, but then the Transient temporarily goes away.

**Comment by tarsius** [comment]
> ```elisp
> (setf transient-display-buffer-action '(display-buffer-same-window))
> ```

That's not a valid configuration. The docstring contains this:

> Note that the buffer that was current before the transient buffer
> is shown should remain the current buffer.  Many suffix commands
> act on the thing at point, if appropriate, and if the transient
> buffer became the current buffer, then that would change what is
> at point.  To that effect `inhibit-same-window' ensures that the
> selected window is not used to show the transient buffer.

**Comment by minad** [comment]
@tarsius I agree that `display-buffer-side-window` is not appropriate. What about the side window configuration, which also seems to create problems?

~~~ elisp
(setf transient-display-buffer-action '(display-buffer-in-side-window
                                         (side . right)))
~~~

**Comment by minad** [comment]
I've already tried to let bind `window-size-fixed` to nil around the `window-resize` call but that didn't help.

https://github.com/minad/vertico/blob/e826dfcb14af5e2cfd88ed110d0208ddc2d37788/vertico.el#L620-L622

**Comment by tarsius** [comment]
> I've already tried to let bind `window-size-fixed` to nil around the `window-resize` call but that didn't help.

That changes the value in vertico's buffer, but the problem is the value in transient's buffer. Anyway, I've now pushed my fix.

**Comment by josephmturner** [comment]
> That changes the value in vertico's buffer, but the problem is the value in transient's buffer. Anyway, I've now pushed my fix. I have verified that that really fixes it.

This fixes the issue on my machine!  Thank you both for the quick response and fix!!

**Comment by josephmturner** [comment]
For future reference, [this is the commit that solves the issue](https://github.com/magit/transient/commit/5478d4e6a73daa6bf0dcaad6cf4767ee4a1589ec#diff-9bf5da9f8d198cfd71b9015366227c09be569aedeaf25b84cc2b729bdf9bba74R3739).

**Comment by aaronjensen** [comment]
Not sure if there's much to do about it here, but for anyone's reference, this change appears to have made `transient-posframe` frames 1 character wide. That library hasn't been touched in years, so maybe it's time to try something else.

**Comment by tarsius** [comment]
There's a pr for that: https://github.com/yanghaoxie/transient-posframe/pull/6.

**Comment by aaronjensen** [comment]
Yeah, that’s where I got the change. I should have linked the pr, thank you.

**Comment by tarsius** [comment]
I've opened a pr with a better fix, https://github.com/yanghaoxie/transient-posframe/pull/7.

**Comment by aaronjensen** [comment]
Wow, that's great, thank you.

**Comment by cyrus-and** [comment]
> `transient--show` sets `window-size-fixed` to `t` in the menu buffer, which is (only) a problem if the window is split horizontally. I'll have to look into how to not do it in that case.
> 
> Probably
> 
> ```diff
> -      (setq window-size-fixed t)
> +      (setq window-size-fixed (if (window-full-height-p) 'width 'both))
> ```

@tarsius this fix introduced in magit/transient@5478d4e6a73daa6bf0dcaad6cf4767ee4a1589ec doesn't play well with [Zoom](https://github.com/cyrus-and/zoom), any chance to make it customizable? Or, are you aware of any workarounds (besides advising and such)?

**Comment by tarsius** [comment]
Please be more precise than "doesn't play well".

Are you sure that using `width` instead of `t` is actually any more problematic for Zoom? Can you explain why?

> any chance to make it customizable?

I am seriously considering it, but to an extend that would mean admitting defeat.

There's a reason Transient fixes the size of its window: we really do want to display all the available bindings, and while fixing the window size does not *absolutely* guarantee that, it very significantly reduces the risk of the window being truncated in cases where there actually would be enough space.

Zoom makes sure the *selected* window is large. That is problematic when it comes to Transient, because that makes sure that the window, which was selected before the transient window is shown, remain the selected window (so that the previous thing-at-point, remains the thing-at-point, while the menu/window is being shown). As a result the transient buffer/window is shown in an area "where the small windows go", which often would mean that some of the available commands cannot be seen.

In other words, it would probably make sense for Zoom to exempt Transient's window from its rule, provided it is flexible enough to support that.

That's what I meant by "admitting defeat" above: it is important that the Transient menu is shown in full, but some packages (*possibly* including Zoom), want to do things that prevent that, and they *may* additionally blow up, if prevented from doing their thing. I *can* deal with that by allowing third-party packages to instruct Transient to sometimes fail to display the full menu, but because the assumption is that the menu is nearly always shown in full (i.e., we don't provide tooling to deal with situations where that is not the case), that would mean that Transient is somewhat broken when used in combination with such a package.

**Comment by cyrus-and** [comment]
> Please be more precise than "doesn't play well".
> 
> Are you sure that using `width` instead of `t` is actually any more problematic for Zoom? Can you explain why?

Sorry, I apologize for the lack of details. So basically using `'width` instead of `t` causes the transient menu to be too small, hence truncating its content.

> In other words, it would probably make sense for Zoom to exempt Transient's window from its rule, provided it is flexible enough to support that.

Zoom is just a glorified `balance-windows` followed by `window-resize`. Ignoring windows, from the Zoom perspective, simply would avoid the latter, so it wouldn't help here, because the transient window is not _zoomed_ anyway.

Instead, fixed-size windows are properly handled, that's why everything used to work nicely before that change.

> That's what I meant by "admitting defeat" above: it is important that the Transient menu is shown in full, but some packages (_possibly_ including Zoom), want to do things that prevent that, and they _may_ additionally blow up, if prevented from doing their thing.

Then I don't see why you just can't fix its size as before. Admittedly, I don't completely understand the issue of the OP.

> I _can_ deal with that by allowing third-party packages to instruct Transient to sometimes fail to display the full menu, but because the assumption is that the menu is nearly always shown in full (i.e., we don't provide tooling to deal with situations where that is not the case), that would mean that Transient is somewhat broken when used in combination with such a package.

If a window is fixed in size, then Zoom never attempt to change it. The only case in which a transient buffer is truncated is when the whole frame is not big enough, but this happens with or without Zoom.

**Comment by tarsius** [comment]
> So basically using `'width` instead of `t` causes the transient menu to be too small, hence truncating its content.

Are you additionally customizing how Transient displays its window?

By default Transient displays its window at the bottom, below some other window. As a result the transient window is not `window-full-height-p` and thus Transient sets the value of `window-size-fixed` to `t`, the same value as before.

**Comment by cyrus-and** [comment]
> Are you additionally customizing how Transient displays its window?

Nope.

> By default Transient displays its window at the bottom, below some other window. As a result the transient window is not `window-full-height-p` and thus Transient sets the value of `window-size-fixed` to `t`, the same value as before.

Well, it doesn't seem to be the case. Manually changing the code back to setting `window-size-fixed` to `t` fixes the issue. Moreover, manually changing the code to print the argument, shows `width`:

```lisp
;; (setq window-size-fixed (if (window-full-height-p) 'width t))
(setq window-size-fixed t)
(print (if (window-full-height-p) 'width t))
```

**Comment by cyrus-and** [comment]
It's easy to reproduce if you feel like, from `emacs -Q` just install Magit and Zoom, then trigger any of the Magit menus.

**Comment by tarsius** [comment]
Ah right. Transient check whether selected window uses the full height, *before* creating its own window.

I'll have to think more about how to consolidate the needs of various possible configurations. It might be a few days until I get to that.

**Comment by tarsius** [comment]
> Then I don't see why you just can't fix its size as before. Admittedly, I don't completely understand the issue of the OP.

It's important to understand that part to be able to reason possible solutions.

We begin with a single magit status buffer and bring up the diff menu. That window is displayed on the left of the status buffer, because of the modification to `transient-display-buffer-action`. We are using vertico or some other framework that displays the completion candidates in the minibuffer. Vertico tries to resize the minibuffer window to make room to display the candidates. To do so other windows have to shrink in size. The only windows that exist are all above the minibuffer window. One of these windows, transients menu, refuses to be resized because of the `(setq window-size-fixed t)`.

> It's easy to reproduce if you feel like, from `emacs -Q` just install Magit and Zoom, then trigger any of the Magit menus.

In my quick test that caused the transient window to needlessly grow, and I was going to say that that isn't so bad and something that is to be expected when using something `balance-windows` on a hook.

But ...

> So basically using `'width` instead of `t` causes the transient menu to be too small, hence truncating its content.

... that is of course more serious.

The matter is complicated because a user may use zoom and vertico, *and* configure transient to display the menu in a window that takes up the full frame height.

In this situation if we use `t`, then the minibuffer cannot grow, but if we use `nil`, then the menu window may shrink causing parts of the menu to be truncated.

So clearly no value for `window-size-fixed` is right at times, even given one particular configuration. (Well, unless using a "recommended configuration" (such as the default), there's a reason the default is the default.)

In fact, transient already temporarily binds `window-size-fixed` to `nil`, when transient refreshes the menu because the user invoked a command and the menu may thus have grown or shrunk (try e.g., `C-x ...` in any menu).

This could be extended to other situations. Transient already does detect when the minibuffer is used, so that could be extended to bind to `nil` in that case too.

But there's always another special case around the next corner. Arbitrary commands can be bound in menus, and some of them are bound to require a new window to be created or some arbitrary window to grow, and we won't be able to anticipate all those cases now. And I do not want to be on the hook to forever deal with additional special cases.

But you can cover your package's use-case in your package, i.e.:

```diff
@@ -213,7 +213,17 @@ (defun zoom--update ()
         ;; siblings
         (window-resize-pixelwise t))
     ;; start from a balanced layout anyway
-    (balance-windows)
+    (if-let ((win (window-live-p (bound-and-true-p transient--window))))
+        (let (old)
+          (unwind-protect
+              (progn
+                (with-current-buffer (window-buffer win)
+                  (setq old window-size-fixed)
+                  (setq window-size-fixed t))
+                (balance-windows))
+            (with-current-buffer (window-buffer win)
+              (setq window-size-fixed old))))
+      (balance-windows))
     ;; check if the selected window is not ignored
     (unless (zoom--window-ignored-p)
       (zoom--resize)
```

**Comment by cyrus-and** [comment]
> It's important to understand that part to be able to reason possible solutions.

Thanks, now I understand the scenario better. Even though, AFAIK, never occurred to me to use `completing-read` (I use vertico too) while a transient buffer is displayed. I don't know, it seems to spoil the _transient_ part of it all...

> But you can cover your package's use-case in your package, i.e.:

It works, but it feels somehow _wrong_ to hardcode a reference to another package instead of relying on just native Emacs features, like fixed-size windows. I mean, what if another package comes along, that requires a similar fix, this is something cumbersome to maintain.

---

You are surely more aware of all the internals and subtleties of this than me, but here's my 2 cents. You (transient developer) cannot really enforce the fact that a transient buffer will always be entirely displayed in a frame, e.g., there might be too many windows, the user might have some large values for `window-min-width` and/or `window-min-height`, the frame might just be too small for the content, add to this all the special cases outlined in this thread...

So it seems that it's up to the user to decide some details about how they want the transient buffer to be displayed, according to their setup.

**Comment by cyrus-and** [comment]
> You probably do this all the time when using magit. For example, when you check out another branch using `b b`, then that uses the minibuffer to read the branch to be checked out.

Well, AFAICT the default behaviour is to hide the transient menu as soon as I type `b b` that's why I never experienced that. And this happens even with the default value of `transient-hide-during-minibuffer-read` which, as you correctly state, is `nil`.

I use Magit every day (BTW, thanks <3) and I can't think of any situation in which I end up with both the transient menu **and** the `completing-read` candidates visible ... but it might be that I'm so accustomed to that, that I don't even notice.

> But ultimately the reason why I think it is on zoom to deal with this incompatibility, is that I feel that it is the "odd one" here. [...]

I get, and kind of agree on, all that. Zoom is a weird beast, as you say implementing a "window manager" is no joke, I spent like maybe a couple of hours trying to implement a Zoom-flavored `balance-windows` from scratch, but then I gave up pretty quickly, due to the complexity and the incredibly high number of corner cases. Instead I opted for the best-effort `balance-windows` + `window-resize` approach.

> I think you should implement a similar "exemption config options" for balance-windows as is already in place for zoom--resize. You could wait to do this until at least a second instance comes along (though given this package's age, I am actually surprised this hasn't happened yet).

I'd be more than happy to add such an exemption mechanism, but I'm not sure it's doable, I need to do some tests, but IIRC there were issues about the initial size of windows, that's why in the [FAQ](https://github.com/cyrus-and/zoom#why-windows-are-resized-even-when-they-are-ignored) I suggest to specify a manual size. Admittedly, this is all fuzzy, I'll come back to you after some experiments... Thanks for your support so far.

PS sorry @minad for hijacking the original issue, albeit related, it's not strictly about vertico, let us know if you prefer we move this elsewhere.

**Comment by tarsius** [comment]
> Well, AFAICT the default behaviour is to hide the transient menu as soon as I type `b b` that's why I never experienced that. And this happens even with the default value of `transient-hide-during-minibuffer-read` which, as you correctly state, is `nil`.

Right, forgot about that distinction: `b b` exits the `b` prefix, in which case the menu is always closed, as soon as the command is invoked; but try `l - -`, an infix argument (which does not exit the menu).

I'm now thinking about enabling `transient-hide-during-minibuffer-read` by default, going back to `(setq window-size-fixed t)`, and adding a warning in the appropriate docstring(s?) about the incompatibility this issue originally was about.

**Comment by cyrus-and** [comment]
> I'm now thinking about enabling `transient-hide-during-minibuffer-read` by default, going back to `(setq window-size-fixed t)`, and adding a warning in the appropriate docstring(s?) about the incompatibility this issue originally was about.

I think it makes perfect sense, as I said, I thought it was the default already.

---

> I'd be more than happy to add such an exemption mechanism, but I'm not sure it's doable, I need to do some tests, but IIRC there were issues about the initial size of windows, that's why in the [FAQ](https://github.com/cyrus-and/zoom#why-windows-are-resized-even-when-they-are-ignored) I suggest to specify a manual size. Admittedly, this is all fuzzy, I'll come back to you after some experiments...

I'm experimenting with this code, based on your good idea of fixing the size before calling `balance-windows`:

https://github.com/cyrus-and/zoom/blob/a02f09be6ce374e9504901edbecfd5697f9ebbf1/zoom.el#L215-L227

It seems to work well so far, at least it seems to fix my original issue with Magit/Transient.

In this case I have `zoom-ignored-buffer-names` set to `'(" *transient*")`.

As I mentioned though, there is a small glitch in some cases, for example, if I ignore `imenu-list-major-mode` (the canonical example of a package that required some manual fix; see the above FAQ) the first time I spawn that window with `imenu-list` its size is slightly larger, then after moving the focus around to other windows it shrinks a bit. I'm not sure what causes this, but still I think it's a good improvement, because users would only need to ignore those windows, instead of fiddling with hooks, as suggested in the FAQ.

**Comment by tarsius** [comment]
I've also tried to address this on my end, see the `fixed` branch. @josephmturner, please give it a try too.

**Comment by cyrus-and** [comment]
OK, that took me a while... the [above patch](https://github.com/minad/vertico/issues/532#issuecomment-2484157001) is wrong, `window-live-p` does not return the window, it returns `t` or `nil`, hence the next `(window-buffer win)` fails, the same as with `(window-buffer t)`, and no `balance-windows` happens. It seemed to work because of that, but indeed it's not a viable solution for transient, sadly.

**Comment by tarsius** [comment]
Well, the fix for that should be simple:

```lisp
(and (window-live-p (bound-and-true-p transient--window))
     transient--window)
```

In your code you should probably use `(window-list (selected-frame))`.

**Comment by cyrus-and** [comment]
> Well, the fix for that should be simple:

No you don't get it. Of course I can work around that. The point is that, if I fix the code (also basically my code attempt that I linked [before](https://github.com/cyrus-and/zoom/blob/a02f09be6ce374e9504901edbecfd5697f9ebbf1/zoom.el#L215-L227)) the result is not what we want, the transient window is truncated. So apparently, when we set the fixed flag to `t` for the transient window, its size is _already wrong_.

This approach might maybe work fine as a better way to ignore certain windows from being balanced in Zoom, but sadly it's not a solution for Magit menus.

> In your code you should probably use (window-list (selected-frame)).

Why? The doc states:

> FRAME nil or omitted means use the selected frame.

**Comment by tarsius** [comment]
> This approach might maybe work fine as a better way to ignore certain windows from being balanced in Zoom, but sadly it's not a solution for Magit menus.

It appears to work for me. Maybe a difference in hook order? This seems like a reasonable and promising approach. I think it's too early to give up on it; its worth looking into how it can be tweaked to always work.

> FRAME nil or omitted means use the selected frame.

Weird I overlooked that. That's how I remembered it, but I somehow didn't see it when I went to double check.

**Comment by cyrus-and** [comment]
> It appears to work for me. Maybe a difference in hook order? This seems like a reasonable and promising approach. I think it's too early to give up on it; its worth looking into how it can be tweaked to always work.

That's strange, it basically never works here. Maybe your frame is big enough to accomodate the _zooming_? Here I'm trying with `emacs -Q`, this [zoom.el](https://github.com/cyrus-and/zoom/blob/window-size-fixed/zoom.el), and:

```lisp
(require 'zoom "/path/to/zoom.el")

(package-initialize) ; to access the already installed Magit

(custom-set-variables
 '(zoom-mode t)
 '(zoom-size '(1.0 . 1.0)))
```

When I say that it doesn't work, I mean that either the Magit menu is truncated or it has too much empty space. That's why I'm using `1.0` above, which means: _try to zoom the window to fit the whole frame_.

> (Is the transient window also truncated for you if `zoom-mode` isn't even enabled?)

Nope, it works as expected in that case.

**Comment by josephmturner** [comment]
> I've also tried to address this on my end, see the `fixed` branch. @josephmturner, please give it a try too.

@tarsius I tried this on Emacs 29.4 in a clean `--init-directory`:

``` elisp
(require 'vertico)
;; Don't use the built-in version.
;; (require 'transient)
(vertico-mode +1)
;; Checked out at `fixed` branch.
(load "/home/joseph/.local/src/transient.el/lisp/transient.el")
;; (setf transient-display-buffer-action '(display-buffer-full-frame))

(transient-define-suffix transient-test-read-minibuffer ()
  :transient t
  (interactive)
  (completing-read "Prompt: " '(a b c d)))

(transient-define-prefix foo ()
  "Test"
  ;; :refresh-suffixes t
  ["Hyperdrive peer graph"
   ("r" "Read" transient-test-read-minibuffer)
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")
   (:info "take up vertical space")]
  (interactive)
  (transient-setup 'foo))
```

I noticed two unusual things.

1. With `transient-hide-during-minibuffer-read` set to `nil` and `transient-display-buffer-action` set to the default, I press `r`, then select any candidate. Then I press `r` again, and this time vertico does not expand and I do not see any candidates (like in the original bug report).

2. With `transient-hide-during-minibuffer-read` set to `t` and `transient-display-buffer-action` set to `'(display-buffer-full-frame)`, I get a `Selecting deleted buffer` error when I attempt to select any of the candidates with `RET`:

<details>
  <summary>Backtrace 1</summary>

```
Debugger entered--Lisp error: (error "Selecting deleted buffer")
  apply(debug
	(error
	 (error "Selecting deleted buffer")))
  transient--exit-and-debug(error
			    (error "Selecting deleted buffer"))
  #<subr completing-read-default>("Prompt: "
				  (a b c d)
				  nil nil nil nil nil nil)
  apply((#<subr completing-read-default> "Prompt: "
		(a b c d)
		nil nil nil nil nil nil))
  vertico--advice(#<subr completing-read-default> "Prompt: "
			 (a b c d)
			 nil nil nil nil nil nil)
  apply(vertico--advice #<subr completing-read-default>
			("Prompt: "
			 (a b c d)
			 nil nil nil nil nil nil))
  completing-read-default("Prompt: "
			  (a b c d)
			  nil nil nil nil nil nil)
  completing-read("Prompt: "
		  (a b c d))
  (lambda nil
    (interactive)
    (completing-read "Prompt: "
		     '(a b c d)))()
  apply((lambda nil
	  (interactive)
	  (completing-read "Prompt: "
			   '(a b c d)))
	nil)
  (let
      ((debugger #'transient--exit-and-debug))
    (apply fn args))
  (unwind-protect
      (let
	  ((debugger #'transient--exit-and-debug))
	(apply fn args))
    (let*
	((unwind
	  (and t
	       (eieio-oref prefix 'unwind-suffix))))
      (if unwind
	  (progn
	    (transient--debug 'unwind-command)
	    (funcall unwind suffix))
	nil))
    (advice-remove suffix advice)
    (eieio-oset prefix 'unwind-suffix nil))
  (closure
      ((advice lambda
	       (fn &rest args)
	       (interactive
		(closure
		    (#3
		     (suffix . transient-test-read-minibuffer)
		     (prefix . #<transient-prefix transient-prefix-d50798>))
		    (spec)
		  (let
		      ((abort t))
		    (unwind-protect
			(prog1
			    (let
				((debugger #'transient--exit-and-debug))
			      (advice-eval-interactive-spec spec))
			  (setq abort nil))
		      (if abort
			  (progn
			    (let*
				((unwind
				  (and t
				       (eieio-oref prefix 'unwind-suffix))))
			      (if unwind
				  (progn
				    (transient--debug 'unwind-interactive)
				    (funcall unwind suffix))
				nil))
			    (advice-remove suffix advice)
			    (eieio-oset prefix 'unwind-suffix nil)))))))
	       (apply '#0 fn args))
       (suffix . transient-test-read-minibuffer)
       (prefix . #<transient-prefix transient-prefix-d50798>))
      (fn &rest args)
    (unwind-protect
	(let
	    ((debugger #'transient--exit-and-debug))
	  (apply fn args))
      (let*
	  ((unwind
	    (and t
		 (eieio-oref prefix 'unwind-suffix))))
	(if unwind
	    (progn
	      (transient--debug 'unwind-command)
	      (funcall unwind suffix))
	  nil))
      (advice-remove suffix advice)
      (eieio-oset prefix 'unwind-suffix nil)))((lambda nil (interactive) (completing-read "Prompt: " '(a b c d))))
  apply((closure
	    ((advice lambda
		     (fn &rest args)
		     (interactive
		      (closure
			  (#4
			   (suffix . transient-test-read-minibuffer)
			   (prefix . #<transient-prefix transient-prefix-d50798>))
			  (spec)
			(let
			    ((abort t))
			  (unwind-protect
			      (prog1
				  (let
				      ((debugger #'transient--exit-and-debug))
				    (advice-eval-interactive-spec spec))
				(setq abort nil))
			    (if abort
				(progn
				  (let*
				      ((unwind
					(and t
					     (eieio-oref prefix 'unwind-suffix))))
				    (if unwind
					(progn
					  (transient--debug 'unwind-interactive)
					  (funcall unwind suffix))
				      nil))
				  (advice-remove suffix advice)
				  (eieio-oset prefix 'unwind-suffix nil)))))))
		     (apply '#1 fn args))
	     (suffix . transient-test-read-minibuffer)
	     (prefix . #<transient-prefix transient-prefix-d50798>))
	    (fn &rest args)
	  (unwind-protect
	      (let
		  ((debugger #'transient--exit-and-debug))
		(apply fn args))
	    (let*
		((unwind
		  (and t
		       (eieio-oref prefix 'unwind-suffix))))
	      (if unwind
		  (progn
		    (transient--debug 'unwind-command)
		    (funcall unwind suffix))
		nil))
	    (advice-remove suffix advice)
	    (eieio-oset prefix 'unwind-suffix nil)))
	(lambda nil
	  (interactive)
	  (completing-read "Prompt: "
			   '(a b c d)))
	nil)
  (lambda
    (fn &rest args)
    (interactive
     (closure
	 ((advice . #0)
	  (suffix . transient-test-read-minibuffer)
	  (prefix . #<transient-prefix transient-prefix-d50798>))
	 (spec)
       (let
	   ((abort t))
	 (unwind-protect
	     (prog1
		 (let
		     ((debugger #'transient--exit-and-debug))
		   (advice-eval-interactive-spec spec))
	       (setq abort nil))
	   (if abort
	       (progn
		 (let*
		     ((unwind
		       (and t
			    (eieio-oref prefix 'unwind-suffix))))
		   (if unwind
		       (progn
			 (transient--debug 'unwind-interactive)
			 (funcall unwind suffix))
		     nil))
		 (advice-remove suffix advice)
		 (eieio-oset prefix 'unwind-suffix nil)))))))
    (apply
     '(closure
	  ((advice . #0)
	   (suffix . transient-test-read-minibuffer)
	   (prefix . #<transient-prefix transient-prefix-d50798>))
	  (fn &rest args)
	(unwind-protect
	    (let
		((debugger #'transient--exit-and-debug))
	      (apply fn args))
	  (let*
	      ((unwind
		(and t
		     (eieio-oref prefix 'unwind-suffix))))
	    (if unwind
		(progn
		  (transient--debug 'unwind-command)
		  (funcall unwind suffix))
	      nil))
	  (advice-remove suffix advice)
	  (eieio-oset prefix 'unwind-suffix nil)))
     fn args))((lambda nil (interactive) (completing-read "Prompt: " '(a b c d))))
  apply((lambda
	  (fn &rest args)
	  (interactive
	   (closure
	       ((advice . #1)
		(suffix . transient-test-read-minibuffer)
		(prefix . #<transient-prefix transient-prefix-d50798>))
	       (spec)
	     (let
		 ((abort t))
	       (unwind-protect
		   (prog1
		       (let
			   ((debugger #'transient--exit-and-debug))
			 (advice-eval-interactive-spec spec))
		     (setq abort nil))
		 (if abort
		     (progn
		       (let*
			   ((unwind
			     (and t
				  (eieio-oref prefix 'unwind-suffix))))
			 (if unwind
			     (progn
			       (transient--debug 'unwind-interactive)
			       (funcall unwind suffix))
			   nil))
		       (advice-remove suffix advice)
		       (eieio-oset prefix 'unwind-suffix nil)))))))
	  (apply
	   '(closure
		((advice . #1)
		 (suffix . transient-test-read-minibuffer)
		 (prefix . #<transient-prefix transient-prefix-d50798>))
		(fn &rest args)
	      (unwind-protect
		  (let
		      ((debugger #'transient--exit-and-debug))
		    (apply fn args))
		(let*
		    ((unwind
		      (and t
			   (eieio-oref prefix 'unwind-suffix))))
		  (if unwind
		      (progn
			(transient--debug 'unwind-command)
			(funcall unwind suffix))
		    nil))
		(advice-remove suffix advice)
		(eieio-oset prefix 'unwind-suffix nil)))
	   fn args))
	(lambda nil
	  (interactive)
	  (completing-read "Prompt: "
			   '(a b c d)))
	nil)
  transient-test-read-minibuffer()
  funcall-interactively(transient-test-read-minibuffer)
  command-execute(transient-test-read-minibuffer)
```

</details>

As soon as a close the backtrace with `q`, another one opens immediately:

<details>
  <summary>Backtrace 2</summary>

```
Debugger entered--Lisp error: (wrong-type-argument
			       (or eieio-object cl-structure-object oclosure)
			       nil)
  apply(debug
	(error
	 (wrong-type-argument
	  (or eieio-object cl-structure-object oclosure)
	  nil)))
  transient--exit-and-debug(error
			    (wrong-type-argument
			     (or eieio-object cl-structure-object oclosure)
			     nil))
  eieio-oref(nil command)
  (eq this-command
      (eieio-oref transient--prefix 'command))
  (and
   (eq this-command
       (eieio-oref transient--prefix 'command))
   (not prefix-arg))
  (cond
   (transient--exitp
    (transient--post-exit))
   ((and
     (eq this-command
	 (eieio-oref transient--prefix 'command))
     (not prefix-arg)))
   (transient--refreshp
    (transient--env-apply #'transient--refresh-transient))
   ((let
	((old transient--redisplay-map)
	 (new
	  (transient--make-redisplay-map)))
      (if
	  (equal old new)
	  nil
	(transient--pop-keymap 'transient--redisplay-map)
	(setq transient--redisplay-map new)
	(transient--push-keymap 'transient--redisplay-map))
      (transient--env-apply #'transient--redisplay))))
  (let
      ((debugger #'transient--exit-and-debug))
    (cond
     (transient--exitp
      (transient--post-exit))
     ((and
       (eq this-command
	   (eieio-oref transient--prefix 'command))
       (not prefix-arg)))
     (transient--refreshp
      (transient--env-apply #'transient--refresh-transient))
     ((let
	  ((old transient--redisplay-map)
	   (new
	    (transient--make-redisplay-map)))
	(if
	    (equal old new)
	    nil
	  (transient--pop-keymap 'transient--redisplay-map)
	  (setq transient--redisplay-map new)
	  (transient--push-keymap 'transient--redisplay-map))
	(transient--env-apply #'transient--redisplay)))))
  (condition-case err
      (let
	  ((debugger #'transient--exit-and-debug))
	(cond
	 (transient--exitp
	  (transient--post-exit))
	 ((and
	   (eq this-command
	       (eieio-oref transient--prefix 'command))
	   (not prefix-arg)))
	 (transient--refreshp
	  (transient--env-apply #'transient--refresh-transient))
	 ((let
	      ((old transient--redisplay-map)
	       (new
		(transient--make-redisplay-map)))
	    (if
		(equal old new)
		nil
	      (transient--pop-keymap 'transient--redisplay-map)
	      (setq transient--redisplay-map new)
	      (transient--push-keymap 'transient--redisplay-map))
	    (transient--env-apply #'transient--redisplay)))))
    ((debug error)
     (transient--emergency-exit :post-command)
     (signal
      (car err)
      (cdr err))))
  (if
      (transient--premature-post-command)
      nil
    (transient--debug 'post-command)
    (condition-case err
	(let
	    ((debugger #'transient--exit-and-debug))
	  (cond
	   (transient--exitp
	    (transient--post-exit))
	   ((and
	     (eq this-command
		 (eieio-oref transient--prefix 'command))
	     (not prefix-arg)))
	   (transient--refreshp
	    (transient--env-apply #'transient--refresh-transient))
	   ((let
		((old transient--redisplay-map)
		 (new
		  (transient--make-redisplay-map)))
	      (if
		  (equal old new)
		  nil
		(transient--pop-keymap 'transient--redisplay-map)
		(setq transient--redisplay-map new)
		(transient--push-keymap 'transient--redisplay-map))
	      (transient--env-apply #'transient--redisplay)))))
      ((debug error)
       (transient--emergency-exit :post-command)
       (signal
	(car err)
	(cdr err))))
    (setq transient-current-prefix nil)
    (setq transient-current-command nil)
    (setq transient-current-suffixes nil))
  transient--post-command()
```

</details>

**Comment by tarsius** [comment]
> I noticed two unusual things.

1. I think that would also have happened before I started to make changes to the value of `window-size-fixed`. I'll investigate later.
2. As I mentioned above `'(display-buffer-full-frame [...])` is invalid here. I have now added a check and if that is used a error saying just that is signaled.

**Comment by josephmturner** [comment]
> > I noticed two unusual things.
> 
>     1. I think that would also have happened before I started to make changes to the value of `window-size-fixed`. I'll investigate later.

Thanks!  I've made a separate issue https://github.com/magit/transient/issues/330 

>     2. As I mentioned above `'(display-buffer-full-frame [...])` is invalid here. I have now added a check and if that is used a error saying just that is signaled.

Oh, right...  Thanks for your catch, and I think the error message is a good idea.
